File: kernel/new_data/gaussian.js
var fs = require("fs");
var encoding = "utf-8";
var resolution = 10;
var start_sigma = 1;
var end_sigma = 1;
var sigma_increment = 1;
var sigma = 1;
/*
var resolution = process.argv[2];
var start_sigma = process.argv[3];
var end_sigma = process.argv[4];
var sigma_increment = process.argv[5];
*/


var filename = "./source_data/r" + resolution + ".json";
var CHR_NUMBER = 16;

var chr_list = [];
for(var i = 1; i <= CHR_NUMBER; i++) {
  chr_list.push(i);
}

if(resolution != null && start_sigma != null && end_sigma != null && sigma_increment) {
  fs.readFile(filename, encoding, function(err, matrix) {
    matrix = JSON.parse(matrix);

    chr_list.forEach(function(source) {
      chr_list.forEach(function(target) {
	var start_row = (target - 1) * resolution;
	var start_col = (source - 1) * resolution;
	var list_filename = "./list/r" + resolution + "_chr" + source + "-" + target + ".json";
	fs.readFile(list_filename, encoding, function(err, list) {
	  list = JSON.parse(list);
	  list.forEach(function(position) {
	    for(var row = start_row; row < start_row + resolution; row++) {
	      for(var col = start_col; col < start_col + resolution; col++) {
		//console.log("before_matrix[" + row + "][" + col + "] = " + matrix[row][col]);
		matrix[row][col] += gaussian(position[0], position[1], col, row, sigma);
		matrix[row][col] = Math.round(10000 * matrix[row][col]) / 10000;
		//console.log("after_matrix[" + row + "][" + col + "] = " + matrix[row][col]);
	      }
	    }
	  });
	  setTimeout(function() {
	    //console.log(matrix);
	    fs.open("./matrix/r" + resolution + "_s" + sigma + ".json", 'w', function(err, fd) {
	      fs.write(fd, "[");
	      matrix.forEach(function(element_row, i) {
		fs.write(fd, "[" + element_row + "]");
		if(i != (CHR_NUMBER * resolution - 1)) {
		  fs.write(fd, ",\n");
		}
	      });
	      fs.write(fd, "\n]");
	    });
	  }, 5000);
	  
	});
      });
    });

  });
}
else {
  console.log("error command");
}


function gaussian(cx, cy, x, y, s) {
  var coefficient = 1 / (s * Math.sqrt(2 * Math.PI));
  var constant = 2 * s * s;
  var distance = Math.pow(cx - x, 2) + Math.pow(cy - y, 2);
  var result = Math.round(10000 * coefficient * Math.exp(-distance / constant)) / 10000;
  //console.log("result = " + result);
  return result;
}

File: kernel/new_data/gaussian.pl
#!/user/bin/perl -w

use strict;
use JSON;
use Math::Trig;

my $line;
my @array;
my $CHR_NUMBER = 16;
my $resolution;
my $start_sigma = 1;
my $end_sigma = 100;
my $increment = 5;
my $sigma;
my $tmp;
my $s;
my $t;
my $i;
my $j;
my $k;
my $index;

my $time1;
my $time2; 
my $timedif;

my $row;
my $col;
my @list;

my $element;
my @cor;
my $st_row;
my $st_col;
my $source;
my $target;

my $tmpp;

my $tmp_x;
my $tmp_y;

for($resolution = 30; $resolution <= 100; $resolution = $resolution + 10) {

for($sigma = $start_sigma; $sigma <= $resolution; $sigma = $sigma + $increment) {
  if($sigma == 6) {
    $sigma = $sigma - 1;
  }
  print "----------------------------------------------------------------\n";
  print "resolution = $resolution, sigma = $sigma\n";
  print "Reading file ...\n";

  open FILE, "< ./source_data/r".$resolution.".json";

  $tmp = "";
  while($line = <FILE>) {
    chomp $line;
    $tmp = $tmp.$line;
  }
  close FILE;
  $tmp =~ s/\[//g;
  $tmp =~ s/\]//g;
  @array = split(",", $tmp);

  $time1 = time();
  print "Read over. Start to gaussian distribution calculating ...\n";

  #&print_matrix;

  for($source = 1; $source <= $CHR_NUMBER; $source++) {
    for($target = 1; $target <= $CHR_NUMBER; $target++) {

      print "resolution = $resolution, sigma = $sigma, source = $source, target = $target\n";
      open FILE, "< ./list/r".$resolution."_chr".$source."-".$target.".json";
      $line = <FILE>;
      if($line) {
	@list = split(",", $line);
	close FILE;

	$st_row = ($target - 1) * $resolution;
	$st_col = ($source - 1) * $resolution;

	while(@list) {
	  $tmp_y = pop @list;
	  $tmp_x = pop @list;
	  for($row = $st_row; $row < $st_row + $resolution; $row++) {
	    for($col = $st_col; $col < $st_col + $resolution; $col++) {
	      $index = &find_index($col, $row);
	      if($tmp_x != $col || $tmp_y != $row) {
		$array[$index] += &gaussian($tmp_x, $tmp_y, $col, $row, $sigma);
	      }
	    }
	  }
	}
      }
    }
  }

  $time2 = time();
  $timedif = $time2 - $time1;
  print "Completed the calculation. Spend $timedif seconds. Now writing to file ...\n";

  #&print_matrix;

  open FILE, "> ./matrix/r".$resolution."_s".$sigma.".json";

  print FILE "[";
  for($i = 0; $i < $resolution * $CHR_NUMBER; $i++) {
    print FILE "[";
    for($j = 0; $j < $resolution * $CHR_NUMBER; $j++) {
      $index = &find_index($j, $i);
      print FILE $array[$index];
      if($j != $resolution * $CHR_NUMBER - 1) {
	print FILE ",";
      }
    }
    print FILE "]";
    if($i != $resolution * $CHR_NUMBER - 1) {
      print FILE ",";
    }
  }
  print FILE "]";

  close FILE;
  print "\n./matrix/r".$resolution."_s".$sigma.".json has been created.\n";


}
}

sub find_index {
  my($col, $row) = @_;
  return ($row * $CHR_NUMBER * $resolution + $col);
}

sub gaussian {
  my($cx, $cy, $x, $y, $sigma) = @_;
  my $coefficient = 1 / ($sigma * sqrt(2 * pi));
  my $constant = 2 * $sigma * $sigma;
  my $distance = ($x - $cx) ** 2 + ($y - $cy) ** 2;
  my $result = int(10000 * $coefficient * exp(-$distance / $constant)) / 10000;
  return $result;
}

sub print_matrix {
  print "\n";
  for($i = 0; $i < $resolution * $CHR_NUMBER; $i++) {
    print "\t".$i;
  }
  for($i = 0; $i < $resolution * $CHR_NUMBER; $i++) {
    print "\n$i";
    for($j = 0; $j < $resolution * $CHR_NUMBER; $j++) {
      $index = &find_index($j, $i);
      print "\t".$array[$index];
    }
  }
  print "\n";
}


